sdk/src/main/java/com/google/cloud/dataflow/sdk/util/common/Counter.java

```diff
@@ -418,17 +461,56 @@ public boolean isCompatibleWith(Counter<?> that) {
 
   //////////////////////////////////////////////////////////////////////////////
 
-  /** The name of this counter. */
-  protected final String name;
-
   /** The kind of aggregation function to apply to this counter. */
   protected final AggregationKind kind;
 
   /** The commit state of this counter. **/
   protected final AtomicReference<CommitState> commitState;
 
+  /**
+   * The full unique name of the counter. Counters must have a
+   * at least a name, and optionally may be associated with
+   * a systemName, an originalName, or both. All of these
+   * names together define a counter uniquely.
+   */
+  @AutoValue
+  public abstract static class Name {
```

Comment:
I'd consider calling this `CounterName` (1) to be closer to then names of classes in Beam and (2) since otherwise `Name` is relatively ambiguous. Fine to leave it as `Name` if the shorter-names are preferred.